File: bot.py
from telegram import Bot, ParseMode
from telegram.error import TelegramError
from datetime import datetime
import schedule
import time
import os
from dotenv import load_dotenv

<<<<<<< HEAD
# Загружаем переменные окружения из файла .env
load_dotenv()

BOT_TOKEN = os.getenv('BOT_TOKEN')
CHAT_ID = os.getenv('CHAT_ID')
=======
BOT_TOKEN = ''
CHAT_ID = ''
>>>>>>> 1b25c823de35aeed521093e969f54a267ff0aac1

if not BOT_TOKEN or not CHAT_ID:
    print('Please provide both BOT_TOKEN and CHAT_ID')
    exit(1)

# Расписание дежурств и замен
duty_schedule = {
    'Monday': {'db_duty': '@MitolKing', 'db_substitute': '@AnbozSultonov', 'frontend_tasks': '@komyobraufzoda', 'support_fixes': '@muhammad0002', 'substitute': '@UmedjonAliev'},
    'Tuesday': {'db_duty': '@AnbozSultonov', 'db_substitute': '@MitolKing', 'frontend_tasks': '@komyobraufzoda', 'support_fixes': '@golibemomov', 'substitute': '@eldor_juraev'},
    'Wednesday': {'db_duty': '@MitolKing', 'db_substitute': '@AnbozSultonov', 'frontend_tasks': '@komyobraufzoda', 'support_fixes': '@UmedjonAliev', 'substitute': '@hafizovabdullo'},
    'Thursday': {'db_duty': '@AnbozSultonov', 'db_substitute': '@MitolKing', 'frontend_tasks': '@komyobraufzoda', 'support_fixes': '@hafizovabdullo', 'substitute': '@muhammad0002'},
    'Friday': {'db_duty': '@MitolKing', 'db_substitute': '@AnbozSultonov', 'frontend_tasks': '@komyobraufzoda', 'support_fixes': '@eldor_juraev', 'substitute': '@AnbozSultonov'}
}

# Инициализация бота
bot = Bot(token=BOT_TOKEN)

# Функция для отправки сообщения о дежурном и замене
def send_duty_message():
    today = datetime.now().strftime('%A')
    schedule_for_today = duty_schedule.get(today)
    
    if schedule_for_today:
        message = (
            f"Сегодняшние дежурные:\n"
            f"— Дежурство по БД: {schedule_for_today['db_duty']}, замена {schedule_for_today['db_substitute']}\n"
            f"— Фронтенд задачи: {schedule_for_today['frontend_tasks']}\n"
            f"— Саппорт/багфиксы: {schedule_for_today['support_fixes']}\n"
            f"Замена задач фронтенда или саппорта при необходимости: {schedule_for_today['substitute']}"
        )
        try:
            sent_message = bot.send_message(chat_id=CHAT_ID, text=message, parse_mode=ParseMode.HTML)
            bot.pin_chat_message(chat_id=CHAT_ID, message_id=sent_message.message_id, disable_notification=True)
            print('Duty message sent and pinned successfully.')
        except TelegramError as e:
            print(f'Error sending duty message: {e}')
    else:
        print(f'No duty schedule for {today}.')

# Планирование задачи на каждый рабочий день в 09:05
for day in ['monday', 'tuesday', 'wednesday', 'thursday', 'friday']:
    getattr(schedule.every(), day).at("16:18").do(send_duty_message)

# Бесконечный цикл для выполнения запланированных задач
print('Bot is running...')
while True:
    schedule.run_pending()
    time.sleep(1)

File: bot.php
<?php

// Получаем токен и chat ID из переменных окружения
$token = getenv('BOT_TOKEN');
$chatId = getenv('CHAT_ID');

// Расписание дежурств и замен
$duty_schedule = [
    'Monday' => ['db_duty' => '@MitolKing', 'db_substitute' => '@AnbozSultonov', 'frontend_tasks' => '@komyobraufzoda', 'support_fixes' => '@muhammad0002', 'substitute' => '@UmedjonAliev'],
    'Tuesday' => ['db_duty' => '@AnbozSultonov', 'db_substitute' => '@MitolKing', 'frontend_tasks' => '@komyobraufzoda', 'support_fixes' => '@golibemomov', 'substitute' => '@eldor_juraev'],
    'Wednesday' => ['db_duty' => '@MitolKing', 'db_substitute' => '@AnbozSultonov', 'frontend_tasks' => '@komyobraufzoda', 'support_fixes' => '@UmedjonAliev', 'substitute' => '@hafizovabdullo'],
    'Thursday' => ['db_duty' => '@AnbozSultonov', 'db_substitute' => '@MitolKing', 'frontend_tasks' => '@komyobraufzoda', 'support_fixes' => '@hafizovabdullo', 'substitute' => '@muhammad0002'],
    'Friday' => ['db_duty' => '@MitolKing', 'db_substitute' => '@AnbozSultonov', 'frontend_tasks' => '@komyobraufzoda', 'support_fixes' => '@eldor_juraev', 'substitute' => '@AnbozSultonov']
];

// Функция для форматирования сообщения
function format_duty_message($schedule_for_today)
{
    return implode("\n", [
        "Сегодняшние дежурные:",
        "- Дежурство по БД: {$schedule_for_today['db_duty']}, замена {$schedule_for_today['db_substitute']}",
        "- Фронтенд задачи: {$schedule_for_today['frontend_tasks']}",
        "- Саппорт/багфиксы: {$schedule_for_today['support_fixes']}",
        "Замена задач фронтенда или саппорта при необходимости: {$schedule_for_today['substitute']}"
    ]);
}

// Функция для отправки сообщения в Telegram
function send_message_to_telegram($token, $chatId, $message)
{
    $url = "https://api.telegram.org/bot$token/sendMessage";
    $data = ['chat_id' => $chatId, 'text' => $message];

    $options = [
        'http' => [
            'header'  => "Content-type: application/x-www-form-urlencoded\r\n",
            'method'  => 'POST',
            'content' => http_build_query($data)
        ]
    ];

    $context  = stream_context_create($options);
    $response = file_get_contents($url, false, $context);

    return json_decode($response, true);
}

// Обновление статуса пользователя
function update_user_status($username, $status, $days)
{
    $absences = load_absences();
    $absences[$username] = [
        'status' => $status,
        'end_date' => date('Y-m-d', strtotime("+$days days"))
    ];
    save_absences($absences);
}

// Загрузка информации об отсутствии из файла
function load_absences()
{
    if (file_exists('absences.json')) {
        $json = file_get_contents('absences.json');
        return json_decode($json, true);
    }
    return [];
}

// Сохранение информации об отсутствии в файл
function save_absences($absences)
{
    file_put_contents('absences.json', json_encode($absences));
}

// Проверка статуса пользователя
function is_user_absent($username)
{
    $absences = load_absences();
    if (isset($absences[$username])) {
        $end_date = $absences[$username]['end_date'];
        if (strtotime($end_date) >= strtotime(date('Y-m-d'))) {
            return true;
        }
    }
    return false;
}

// Обработка входящих сообщений
function handle_incoming_message($message, $duty_schedule)
{
    global $token, $chatId;
    
    if (strpos($message['text'], '/change') === 0) {
        $parts = explode(' ', $message['text']);
        if (count($parts) >= 4) {
            $username = $parts[1];
            $status = $parts[2];
            $days = intval($parts[3]);
            
            // Обновить информацию о статусе пользователя
            update_user_status($username, $status, $days);
            
            $response_text = "Информация о пользователе {$username} обновлена: {$status} на {$days} дней.";
            send_message_to_telegram($token, $chatId, $response_text);
        } else {
            $response_text = "Неверный формат команды. Используйте: /change @username статус дни";
            send_message_to_telegram($token, $chatId, $response_text);
        }
    }
}

// Основная функция для отправки сообщения о дежурных
function send_duty_message($token, $chatId, $duty_schedule)
{
    $today = date('l');
    $schedule_for_today = $duty_schedule[$today] ?? null;

    if ($schedule_for_today) {
        foreach ($schedule_for_today as $role => $user) {
            if (is_user_absent($user)) {
                $schedule_for_today[$role] = $schedule_for_today[$role . '_substitute'] ?? $user;
            }
        }

        $message = format_duty_message($schedule_for_today);
        $response = send_message_to_telegram($token, $chatId, $message);

        if ($response && $response['ok']) {
            echo "Сообщение успешно отправлено в Telegram.";
        } else {
            echo "Ошибка при отправке сообщения в Telegram.";
        }
    } else {
        echo "На сегодня расписание дежурств не найдено.";
    }
}

// Загружаем входящие сообщения
function get_updates($token, $offset)
{
    $url = "https://api.telegram.org/bot$token/getUpdates?offset=$offset";
    $response = file_get_contents($url);
    return json_decode($response, true);
}

// Основная функция
function main()
{
    global $token, $chatId, $duty_schedule;

    $offset = 0;
    while (true) {
        $updates = get_updates($token, $offset);
        if ($updates['ok'] && count($updates['result']) > 0) {
            foreach ($updates['result'] as $update) {
                $offset = $update['update_id'] + 1;
                if (isset($update['message'])) {
                    handle_incoming_message($update['message'], $duty_schedule);
                }
            }
        }
        sleep(1); // чтобы не перегружать сервер Telegram
    }
}

main();
